IntegrationTests/UITests/UITest.cs

```diff
@@ -1,58 +1,181 @@
+                    string formState = form.Visible ? "open" : "exists";
+                    LoggingService.Log($"{MethodBase.GetCurrentMethod().Name} {typeof(T).FullName} still {formState}");
+
+                    CloseOrDispose(form, $"{MethodBase.GetCurrentMethod().Name} {typeof(T).FullName}");
+                }
+
+                Assert.IsTrue(Application.OpenForms.Count == 0, $"{Application.OpenForms.Count} open form(s)");
+            }
 
-            await idleCompletionSource.Task;
-            Application.Idle -= HandleApplicationIdle;
+            return;
 
             void HandleApplicationIdle(object sender, EventArgs e)
             {
+                LoggingService.Log($"{MethodBase.GetCurrentMethod().Name}");
                 idleCompletionSource.TrySetResult(default);
             }
         }
 
-        public static void RunForm<T>(
-            Action showDialog,
-            Func<T, Task> runAsync)
-            where T : Form
+        private static void CloseOrDispose(Form form, string description)
         {
-            // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
-            // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
-            // complete.
-            var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
+            try
             {
-                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-                await WaitForIdleAsync();
-                var dialog = Application.OpenForms.OfType<T>().Single();
+                if (form == null)
+                {
+                    LoggingService.Log($"{description} form is null");
+                    return;
+                }
+
+                var actionDescription = description + " form ";
                 try
                 {
-                    await runAsync(dialog);
+                    if (form.Visible)
+                    {
+                        actionDescription += "close";
+                        form.Close(); // also calls form.Dispose()
+                    }
+                    else
+                    {
+                        actionDescription += "dispose";
+                        form.Dispose();
+                    }
+
+                    LoggingService.Log($"{actionDescription}d");
                 }
-                finally
+                catch (Exception ex)
                 {
-                    dialog.Close();
+                    LoggingService.Log($"{actionDescription} threw {ex.Demystify()}");
+                    form.Dispose();
+                    LoggingService.Log($"{description} form disposed");
                 }
-            });
-
-            showDialog();
+            }
+            finally
+            {
+                // Wait for potential pending asynchronous tasks triggered by the form.
+                WaitForPendingOperations();
```

Comment:
📝 This is already going to happen during test cleanup in `ConfigureJoinableTaskFactoryAttribute.AfterTest`. There is no need to duplicate it here.